### PR TITLE
Auto-magical whitespace

### PR DESCRIPTION
**Pull request by matklad**
bors r+
🤖

### CODE REVIEW

**Comment by matklad** [comment]
changelog: internal API for structural editing can automatically insert whitespace in simple cases.

File: crates/syntax/src/ast/edit_in_place.rs [hunk 1]
```diff
@@ -93,21 +93,16 @@ impl GenericParamsOwnerEdit for ast::Enum {
...
-        elements.insert(0, ws)
-    } else {
-        elements.push(ws)
-    }
-    ted::insert_all(position, elements);
+fn create_where_clause(position: Position) {
+    let where_clause: SyntaxElement =
+        make::where_clause(empty()).clone_for_update().syntax().clone().into();
+    ted::insert_ws(position, where_clause);
 }
```
Comment (matklad) [inline crates/syntax/src/ast/edit_in_place.rs:106]: I've been dreaming about the API which can magically fix up whitespace for years, super satisfying to just finally have it!

**Comment by matklad** [comment]
@Veykril medium term goal here would be to tear down the syntax rewriter :0)

**Comment by Veykril** [comment]
> @Veykril medium term goal here would be to tear down the syntax rewriter :0)

That was my immediate thought after I saw this PR 😄

File: crates/syntax/src/ast/edit_in_place.rs (hunk 1 above)
Comment (SomeoneToIgnore) [reply crates/syntax/src/ast/edit_in_place.rs:106]: What about matching brackets insertion, would that be possible to have?

**Comment by matklad** [comment]
I think it’s better to keep this restricted to whitespace only. For
significant tokens, like braces and commas, we should have structured
editing api (edit_in_place at the moment)

**Pull request merged**